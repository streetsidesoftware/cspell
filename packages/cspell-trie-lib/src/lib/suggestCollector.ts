import { createTimer } from './timer';
import { JOIN_SEPARATOR, WORD_SEPARATOR } from './walker';

const defaultMaxNumberSuggestions = 10;

const baseCost = 100;
const maxNumChanges = 5;
const maxCostScale = 0.5;
// max allowed cost scale should be a bit over 50% to allow for suggestions to short words, but not too high to have too many suggestions.
const maxAllowedCostScale = 1.03 * maxCostScale;

const collator = new Intl.Collator();

const regexSeparator = new RegExp(`[${regexQuote(JOIN_SEPARATOR + WORD_SEPARATOR)}]`, 'g');

const wordLengthCost = [0, 50, 25, 5, 0];
const extraWordsCost = 5;

/** time in ms */
const defaultCollectorTimeout = 1000;

export type Cost = number;
export type MaxCost = Cost;

export interface SuggestionResult {
    word: string;
    cost: Cost;
}

export interface Progress {
    type: 'progress';
    /** Number of Completed Tasks so far */
    completed: number;
    /**
     * Number of tasks remaining, this number is allowed to increase over time since
     * completed tasks can generate new tasks.
     */
    remaining: number;
}

const symStopProcessing = Symbol('Collector Stop Processing');

export type GenerateNextParam = MaxCost | symbol | undefined;
export type GenerateSuggestionResult = SuggestionResult | Progress | undefined;

/**
 * Ask for the next result.
 * maxCost - sets the max cost for following suggestions
 * This is used to limit which suggestions are emitted.
 * If the `iterator.next()` returns `undefined`, it is to request a value for maxCost.
 *
 * The SuggestionIterator is generally the
 */
export type SuggestionGenerator = Generator<GenerateSuggestionResult, void, GenerateNextParam>;

// comparison function for Suggestion Results.
export function compSuggestionResults(a: SuggestionResult, b: SuggestionResult): number {
    return a.cost - b.cost || a.word.length - b.word.length || collator.compare(a.word, b.word);
}

export interface SuggestionCollector {
    /**
     * Collection suggestions from a SuggestionIterator
     * @param src - the SuggestionIterator used to generate suggestions.
     * @param timeout - the amount of time in milliseconds to allow for suggestions.
     * before sending `symbolStopProcessing`
     * Iterator implementation:
     * @example
     * r = yield(suggestion);
     * if (r === collector.symbolStopProcessing) // ...stop generating suggestions.
     */
    collect: (src: SuggestionGenerator, timeout?: number) => void;
    add: (suggestion: SuggestionResult) => SuggestionCollector;
    readonly suggestions: SuggestionResult[];
    readonly maxNumChanges: number;
    readonly maxCost: number;
    readonly word: string;
    readonly maxNumSuggestions: number;
    readonly includesTies: boolean;
    readonly ignoreCase: boolean;
    /**
     * Possible value sent to the SuggestionIterator telling it to stop processing.
     */
    readonly symbolStopProcessing: symbol;
}

export interface SuggestionCollectorOptions {
    /**
     * number of best matching suggestions.
     */
    numSuggestions: number;

    /**
     * An optional filter function that can be used to limit remove unwanted suggestions.
     * I.E. to remove forbidden terms.
     */
    filter?: (word: string, cost: number) => boolean;

    /**
     * The number of letters that can be changed when looking for a match
     */
    changeLimit: number | undefined;

    /**
     * Include suggestions with tied cost even if the number is greater than `numSuggestions`.
     */
    includeTies?: boolean;

    /**
     * specify if case / accents should be ignored when looking for suggestions.
     */
    ignoreCase: boolean | undefined;
}

export const defaultSuggestionCollectorOptions: SuggestionCollectorOptions = {
    numSuggestions: defaultMaxNumberSuggestions,
    filter: () => true,
    changeLimit: maxNumChanges,
    includeTies: true,
    ignoreCase: true,
};

export function suggestionCollector(wordToMatch: string, options: SuggestionCollectorOptions): SuggestionCollector {
    const { filter = () => true, changeLimit = maxNumChanges, includeTies = false, ignoreCase = true } = options;
    const numSuggestions = Math.max(options.numSuggestions, 0) || 0;
    const sugs = new Map<string, SuggestionResult>();
    let maxCost: number = baseCost * Math.min(wordToMatch.length * maxAllowedCostScale, changeLimit);

    function dropMax() {
        if (sugs.size < 2) {
            sugs.clear();
            return;
        }
        const sorted = [...sugs.values()].sort(compSuggestionResults);
        let i = numSuggestions - 1;
        maxCost = sorted[i].cost;
        for (; i < sorted.length && sorted[i].cost <= maxCost; ++i) {
            /* empty */
        }
        for (; i < sorted.length; ++i) {
            sugs.delete(sorted[i].word);
        }
    }

    function adjustCost(sug: SuggestionResult): SuggestionResult {
        const words = sug.word.split(regexSeparator);
        const extraCost =
            words.map((w) => wordLengthCost[w.length] || 0).reduce((a, b) => a + b, 0) +
            (words.length - 1) * extraWordsCost;
        return { word: sug.word, cost: sug.cost + extraCost };
    }

    function handleProgress(_progress: Progress) {
        // Do nothing.
    }

    function collectSuggestion(suggestion: SuggestionResult): MaxCost {
        const { word, cost } = adjustCost(suggestion);
        if (cost <= maxCost && filter(suggestion.word, cost)) {
            const known = sugs.get(word);
            if (known) {
                known.cost = Math.min(known.cost, cost);
            } else {
                sugs.set(word, { word, cost });
                if (cost < maxCost && sugs.size > numSuggestions) {
                    dropMax();
                }
            }
        }
        return maxCost;
    }

    /**
     * Collection suggestions from a SuggestionIterator
     * @param src - the SuggestionIterator used to generate suggestions.
     * @param timeout - the amount of time in milliseconds to allow for suggestions.
     */
    function collect(src: SuggestionGenerator, timeout = defaultCollectorTimeout) {
        let stop: false | symbol = false;
        const timer = createTimer();

        let ir: IteratorResult<SuggestionResult | Progress | undefined>;
        while (!(ir = src.next(stop || maxCost)).done) {
            if (timer.elapsed() > timeout) {
                stop = symStopProcessing;
            }
            const { value } = ir;
            if (!value) continue;
            if (isSuggestionResult(value)) {
                collectSuggestion(value);
                continue;
            }
            handleProgress(value);
        }
    }

    function suggestions() {
        const sorted = [...sugs.values()].sort(compSuggestionResults);
        if (!includeTies && sorted.length > numSuggestions) {
            sorted.length = numSuggestions;
        }
        return sorted;
    }

    const collector: SuggestionCollector = {
        collect,
        add: function (suggestion: SuggestionResult) {
            collectSuggestion(suggestion);
            return this;
        },
        get suggestions() {
            return suggestions();
        },
        get maxCost() {
            return maxCost;
        },
        get word() {
            return wordToMatch;
        },
        get maxNumSuggestions() {
            return numSuggestions;
        },
        get maxNumChanges() {
            return changeLimit;
        },
        includesTies: includeTies,
        ignoreCase,
        symbolStopProcessing: symStopProcessing,
    };

    return collector;
}

export function isSuggestionResult(s: GenerateSuggestionResult): s is SuggestionResult {
    const r = s as Partial<SuggestionResult> | undefined;
    return r?.cost !== undefined && r.word != undefined;
}

/**
 *
 * @param text verbatim text to be inserted into a regexp
 * @returns text that can be used in a regexp.
 */
function regexQuote(text: string): string {
    return text.replace(/[[\]\-+(){},|*.\\]/g, '\\$1');
}
