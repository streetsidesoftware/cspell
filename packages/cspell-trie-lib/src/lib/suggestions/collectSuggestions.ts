import type { Progress, SuggestionGenerator, SuggestionResult } from './SuggestionTypes.js';

function isProgress(v: unknown | Progress): v is Progress {
    if (!v || typeof v !== 'object') return false;
    return (v as Progress).type === 'progress';
}

export function collectSuggestions(sugGen: SuggestionGenerator, maxCost = 300, numSugs = 10): SuggestionResult[] {
    const sugs = new Set<SuggestionResult>();
    const sugMap = new Map<string, SuggestionResult>();

    for (let n = sugGen.next(maxCost); !n.done; n = sugGen.next(maxCost)) {
        const sug = n.value;
        if (!sug || isProgress(sug) || sug.cost > maxCost) continue;
        const pref = sugMap.get(sug.word);
        if (pref && pref.cost <= sug.cost) continue;
        if (pref) {
            sugs.delete(pref);
        }
        sugMap.set(sug.word, sug);
        sugs.add(sug);

        if (sugs.size > numSugs) {
            const sorted = [...sugs].sort(compare);
            maxCost = sorted.slice(0, -1).reduce((a, b) => Math.max(a, b.cost), 0);
            sugs.delete(sorted[sorted.length - 1]);
        }
    }

    return [...sugs].sort(compare);

    function compare(a: SuggestionResult, b: SuggestionResult): number {
        return a.cost - b.cost;
    }
}
