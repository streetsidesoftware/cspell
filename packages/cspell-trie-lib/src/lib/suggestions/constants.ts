export const DEFAULT_COMPOUNDED_WORD_SEPARATOR = 'âˆ™';
