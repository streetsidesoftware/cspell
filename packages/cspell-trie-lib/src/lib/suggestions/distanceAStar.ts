import { PairingHeap } from '../utils/PairingHeap';

/**
 * Calculate the edit distance between two words using an A* algorithm.
 *
 * Using basic weights, this algorithm has the same results as the Damerau-Levenshtein algorithm.
 */
export function distanceAStar(a: string, b: string): number {
    const aN = a.length;
    const bN = b.length;
    const cost = 100;

    const candidates = new PairingHeap(compare);

    candidates.add({ ai: 0, bi: 0, c: 0 });

    function opSub(n: Node) {
        const { ai, bi, c } = n;
        if (ai < aN && bi < bN) {
            const cc = a[ai] === b[bi] ? c : c + cost;
            candidates.add({ ai: ai + 1, bi: bi + 1, c: cc });
        }
    }

    function opIns(n: Node) {
        const { ai, bi, c } = n;
        if (bi < bN) {
            candidates.add({ ai: ai, bi: bi + 1, c: c + cost });
        }
    }

    function opDel(n: Node) {
        const { ai, bi, c } = n;
        if (ai < aN) {
            candidates.add({ ai: ai + 1, bi: bi, c: c + cost });
        }
    }

    function opSwap(n: Node) {
        const { ai, bi, c } = n;
        if (a[ai] === b[bi + 1] && a[ai + 1] === b[bi]) {
            candidates.add({ ai: ai + 2, bi: bi + 2, c: c + cost });
        }
    }

    let best: Node | undefined;
    // const bc2 = 2 * bc;
    while ((best = candidates.dequeue())) {
        if (best.ai === aN && best.bi === bN) break;

        opSwap(best);
        opIns(best);
        opDel(best);
        opSub(best);
    }

    return best?.c ?? -1;
}

interface Pos {
    /** the offset in string `a` */
    ai: number;
    /** the offset in string `b` */
    bi: number;
}

interface Node extends Pos {
    /** the current cost */
    c: number;
}

function compare(a: Node, b: Node): number {
    // Choose lowest cost or farthest Manhattan distance.
    return a.c - b.c || b.ai + b.bi - a.ai - a.bi;
}
