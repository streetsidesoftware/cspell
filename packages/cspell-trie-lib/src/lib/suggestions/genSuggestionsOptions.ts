import { WeightMap } from '..';
import { CompoundWordsMethod } from './walker';

export interface GenSuggestionOptionsStrict {
    /**
     * Controls forcing compound words.
     * @default CompoundWordsMethod.NONE
     */
    compoundMethod?: CompoundWordsMethod;

    /**
     * ignore case when searching.
     */
    ignoreCase: boolean;

    /**
     * Maximum number of "edits" allowed.
     * 3 is a good number. Above 5 can be very slow.
     */
    changeLimit: number;

    /**
     * Inserts a compound character between compounded word segments.
     * @default ""
     */
    compoundSeparator?: string;
}

export type GenSuggestionOptions = Partial<GenSuggestionOptionsStrict>;

export interface SuggestionOptionsStrict extends GenSuggestionOptionsStrict {
    /**
     * Maximum number of suggestions to make.
     */
    numSuggestions: number;

    /**
     * Allow ties when making suggestions.
     * if `true` it is possible to have more than `numSuggestions`.
     */
    includeTies: boolean;

    /**
     * Time alloted in milliseconds to generate suggestions.
     */
    timeout: number;

    /**
     * Optional filter function.
     * return true to keep the candidate.
     */
    filter?: (word: string, cost: number) => boolean;

    /**
     * Apply weights to improve the suggestions.
     */
    weightMap?: WeightMap;
}

export type SuggestionOptions = Partial<SuggestionOptionsStrict>;

export const defaultGenSuggestionOptions: GenSuggestionOptionsStrict = {
    compoundMethod: CompoundWordsMethod.NONE,
    ignoreCase: true,
    changeLimit: 5,
};

export const defaultSuggestionOptions: SuggestionOptionsStrict = {
    ...defaultGenSuggestionOptions,
    numSuggestions: 8,
    includeTies: true,
    timeout: 5000,
};

type KeyOfGenSuggestionOptionsStrict = keyof GenSuggestionOptionsStrict;

type KeyMapOfGenSuggestionOptionsStrict = {
    [K in KeyOfGenSuggestionOptionsStrict]: K;
};

type KeyOfSuggestionOptionsStrict = keyof SuggestionOptionsStrict;

type KeyMapOfSuggestionOptionsStrict = {
    [K in KeyOfSuggestionOptionsStrict]: K;
};

const keyMapOfGenSuggestionOptionsStrict: KeyMapOfGenSuggestionOptionsStrict = {
    changeLimit: 'changeLimit',
    compoundMethod: 'compoundMethod',
    ignoreCase: 'ignoreCase',
    compoundSeparator: 'compoundSeparator',
} as const;

const keyMapOfSuggestionOptionsStrict: KeyMapOfSuggestionOptionsStrict = {
    ...keyMapOfGenSuggestionOptionsStrict,
    filter: 'filter',
    includeTies: 'includeTies',
    numSuggestions: 'numSuggestions',
    timeout: 'timeout',
    weightMap: 'weightMap',
};

/**
 * Create suggestion options using composition.
 * @param opts - partial options.
 * @returns Options - with defaults.
 */
export function createSuggestionOptions(...opts: SuggestionOptions[]): SuggestionOptionsStrict {
    const options = { ...defaultSuggestionOptions };
    const keys = Object.keys(keyMapOfSuggestionOptionsStrict) as (keyof SuggestionOptions)[];
    for (const opt of opts) {
        for (const key of keys) {
            assign(options, opt, key);
        }
    }
    return options;
}

function assign<T, K extends keyof T>(dest: T, src: Partial<T>, k: K) {
    dest[k] = src[k] ?? dest[k];
}
