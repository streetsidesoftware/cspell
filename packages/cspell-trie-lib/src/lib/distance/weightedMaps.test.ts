import type { SuggestionCostMapDef } from '../models/suggestionCostsDef';
import { DEFAULT_COMPOUNDED_WORD_SEPARATOR } from '../suggestions/suggestCollector';
import {
    addAdjustment,
    addDefToWeightMap,
    CostPosition,
    createWeightMap,
    lookupReplaceCost,
    PenaltyAdjustment,
    prettyPrintWeightMap,
    __testing__,
} from './weightedMaps';

const { splitMapSubstrings, splitMap, findTrieCostPrefixes, normalizeDef } = __testing__;

const oc = expect.objectContaining;

// const u = undefined;  cspell:

describe('Validate weightedMaps', () => {
    test.each`
        map             | expected
        ${''}           | ${[]}
        ${'abc'}        | ${'abc'.split('')}
        ${'f(ph)(gh)v'} | ${['f', 'ph', 'gh', 'v']}
    `('splitMapSubstrings "$map"', ({ map, expected }) => {
        expect(splitMapSubstrings(map)).toEqual(expected);
    });

    test.each`
        map                 | expected
        ${''}               | ${[]}
        ${'||'}             | ${[]}
        ${'abc'}            | ${['abc'.split('')]}
        ${'abc|'}           | ${['abc'.split('')]}
        ${'f(ph)(gh)v|eé'}  | ${[['f', 'ph', 'gh', 'v'], ['e', 'é', 'é'.normalize('NFD')]]}
        ${'f(ph)(😁)🤣|eé'} | ${[['f', 'ph', '😁', '🤣'], ['e', 'é', 'é'.normalize('NFD')]]}
    `('splitMap "$map"', ({ map, expected }) => {
        expect(splitMap({ map })).toEqual(expected);
    });

    test.each`
        defs                                  | expected
        ${[]}                                 | ${oc({ insDel: {}, replace: {}, swap: {} })}
        ${[defIns('ab', 3)]}                  | ${oc({ insDel: { n: { a: { c: 3 }, b: { c: 3 } } }, replace: {}, swap: {} })}
        ${[defIns('ab', 3), defIns('bc', 2)]} | ${oc({ insDel: { n: { a: { c: 3 }, b: { c: 2 }, c: { c: 2 } } }, replace: {}, swap: {} })}
        ${[defRep('ab', 3)]}                  | ${oc({ insDel: {}, replace: { n: { a: { t: { n: { b: { c: 3 } } } }, b: { t: { n: { a: { c: 3 } } } } } }, swap: {} })}
        ${[defSwap('ab', 3)]}                 | ${oc({ insDel: {}, replace: {}, swap: { n: { a: { t: { n: { b: { c: 3 } } } }, b: { t: { n: { a: { c: 3 } } } } } } })}
    `('buildWeightMap $defs', ({ defs, expected }) => {
        expect(createWeightMap(...defs)).toEqual(expected);
    });

    test.each`
        def1                                               | def2                                       | def3
        ${undefined}                                       | ${undefined}                               | ${undefined}
        ${defIns('ab', 3)}                                 | ${undefined}                               | ${undefined}
        ${defIns('ab', 3)}                                 | ${defIns('bc', 2)}                         | ${undefined}
        ${defRep('ab', 3)}                                 | ${undefined}                               | ${undefined}
        ${defSwap('ab', 3)}                                | ${undefined}                               | ${undefined}
        ${defRep('(ei)(ie)', 3)}                           | ${undefined}                               | ${undefined}
        ${defRep('(ei)(ie)e', 3, { insDel: 11, swap: 4 })} | ${{ map: 'aeiou', replace: 4, insDel: 7 }} | ${defRep('eio', 1, { swap: 3 })}
    `('buildWeightMap pp $def1, $def2, $def2', ({ def1, def2, def3 }) => {
        const defs = [def1, def2, def3].filter((a) => !!a);
        const pp = prettyPrintWeightMap(createWeightMap(...defs));
        expect(pp).toMatchSnapshot();

        const map2 = defs.reverse().reduce((map, def) => addDefToWeightMap(map, def), createWeightMap());
        expect(prettyPrintWeightMap(map2)).toBe(pp);
    });

    test.each`
        defs                                                                    | word      | offset | expected
        ${[]}                                                                   | ${''}     | ${1}   | ${[]}
        ${[defIns('aeiou', 7, penalty(5)), defIns('(ae)(ea)(ou)(ei)(ie)', 11)]} | ${'read'} | ${1}   | ${[{ i: 2, c: 7, p: 5 }, { i: 3, c: 11, p: 0 }] /* cspell:disable-line */}
        ${[defIns('aeiou', 7, penalty(5)), defIns('(ae)(ea)(ou)(ei)(ie)', 11)]} | ${'read'} | ${2}   | ${[{ i: 3, c: 7, p: 5 }] /* cspell:disable-line */}
    `('findTrieCostPrefixes with insDel $defs.0 $defs.1 $word $offset', ({ defs, word, offset, expected }) => {
        const map = createWeightMap(...defs);
        expect([...findTrieCostPrefixes(map.insDel, word, offset)]).toEqual(expected);
    });

    // cspell:ignore aeiou
    test.each`
        defs                                                        | wordA     | ai   | wordB      | bi   | expected
        ${[]}                                                       | ${''}     | ${1} | ${''}      | ${0} | ${[]}
        ${[defIns('aeiou', 7), defIns('(ae)(ea)(ou)(ei)(ie)', 11)]} | ${'read'} | ${1} | ${''}      | ${0} | ${[{ ai: 2, bi: 0, c: 1007, p: 1000 }, { ai: 3, bi: 0, c: 1011, p: 1000 }]}
        ${[defIns('aeiou', 7), defIns('(ae)(ea)(ou)(ei)(ie)', 11)]} | ${'read'} | ${1} | ${'ride'}  | ${1} | ${[{ ai: 2, bi: 1, c: 1007 /* del e */, p: 1000 }, { ai: 3, bi: 1, c: 1011 /* del ea */, p: 1000 }, { ai: 1, bi: 2, c: 1007 /* ins i */, p: 1000 }]}
        ${[defIns('aeiou', 7), defIns('(ae)(ea)(ou)(ei)(ie)', 11)]} | ${'red'}  | ${1} | ${'read'}  | ${1} | ${[{ ai: 2, bi: 1, c: 1007 /* del e */, p: 1000 }, { ai: 1, bi: 2, c: 1007 /* ins */, p: 1000 }, { ai: 1, bi: 3, c: 1011 /* ins ea */, p: 1000 }]}
        ${[defIns('aeiou', 7), defIns('(ae)(ea)(ou)(ei)(ie)', 11)]} | ${'red'}  | ${2} | ${'read'}  | ${2} | ${[{ ai: 2, bi: 3, c: 1007 /* ins a */, p: 1000 }]}
        ${[defIns('1234567890', 7, penalty(20))]}                   | ${'cost'} | ${4} | ${'cost8'} | ${4} | ${[{ ai: 4, bi: 5, c: 1007 /* insert 8 */, p: 1020 }]}
    `(
        'calcInsDelCosts with $defs.0 $defs.1 "$wordA"@$ai, "$wordB"@$bi',
        ({
            defs,
            wordA,
            ai,
            wordB,
            bi,
            expected,
        }: {
            defs: SuggestionCostMapDef[];
            wordA: string;
            wordB: string;
            ai: number;
            bi: number;
            expected: CostPosition[];
        }) => {
            const map = createWeightMap(...defs);
            const results = [...map.calcInsDelCosts({ a: wordA, b: wordB, ai, bi, c: 1000, p: 1000 })];
            expected.forEach((p) => {
                (p.a = p.a ?? wordA), (p.b = p.b ?? wordB);
            });
            expect(results).toEqual(expect.arrayContaining(expected));
            expect(results).toHaveLength(expected.length);
        }
    );

    test.each`
        defs                                                | wordA        | ai   | wordB        | bi   | expected
        ${[]}                                               | ${''}        | ${0} | ${''}        | ${0} | ${[]}
        ${[defRep('aeiou', 5), defRep('ae(ae)(ea)', 7)]}    | ${'read'}    | ${1} | ${'red'}     | ${1} | ${[{ ai: 3, bi: 2, c: 1007 /* ea -> a */, p: 1000 }]}
        ${[defRep('aeiou', 5), defRep('ae(ae)(ea)', 7)]}    | ${'read'}    | ${1} | ${'road'}    | ${1} | ${[{ ai: 2, bi: 2, c: 1005 /* e -> o */, p: 1000 }]}
        ${[defRep('aeiou', 5), defRep('o(oo)|e(ee)', 7)]}   | ${'met'}     | ${1} | ${'meet'}    | ${1} | ${[{ ai: 2, bi: 3, c: 1007 /* e -> ee */, p: 1000 }]}
        ${[defRep('aeiou', 5), defRep('o(oo)|e(ee)', 7)]}   | ${'meet'}    | ${1} | ${'met'}     | ${1} | ${[{ ai: 3, bi: 2, c: 1007 /* ee -> e */, p: 1000 }]}
        ${[defRep('aeiou', 5), defRep('(ei)(ie)', 7)]}      | ${'believe'} | ${3} | ${'receive'} | ${3} | ${[{ ai: 4, bi: 4, c: 1005 /* i => e */, p: 1000 }, { ai: 5, bi: 5, c: 1007 /* ie => ei */, p: 1000 }]}
        ${[defRep('(sk)(sch)', 5), defRep('(sch)(sk)', 7)]} | ${'school'}  | ${0} | ${'skull'}   | ${0} | ${[{ ai: 3, bi: 2, c: 1005 /* sch => sk */, p: 1000 }]}
    `(
        'calcReplaceCosts with $defs.0 $defs.1 "$wordA"@$ai, "$wordB"@$bi',
        ({
            defs,
            wordA,
            ai,
            wordB,
            bi,
            expected,
        }: {
            defs: SuggestionCostMapDef[];
            wordA: string;
            wordB: string;
            ai: number;
            bi: number;
            expected: CostPosition[];
        }) => {
            const map = createWeightMap(...defs);
            const results = [...map.calcReplaceCosts({ a: wordA, b: wordB, ai, bi, c: 1000, p: 1000 })];
            expected.forEach((p) => {
                (p.a = p.a ?? wordA), (p.b = p.b ?? wordB);
            });
            expect(results).toEqual(expect.arrayContaining(expected));
            expect(results).toHaveLength(expected.length);
        }
    );

    test.each`
        defs                                    | wordA        | ai   | wordB        | bi   | expected
        ${[defSwap('ae', 9), defSwap('ei', 7)]} | ${'believe'} | ${1} | ${'receive'} | ${1} | ${[]}
        ${[defSwap('ae', 9), defSwap('ei', 7)]} | ${'believe'} | ${3} | ${'receive'} | ${3} | ${[{ ai: 5, bi: 5, c: 1007 /* swap ei -> ie */, p: 1000 }]}
        ${[defSwap('ei', 7, penalty(20))]}      | ${'believe'} | ${3} | ${'receive'} | ${3} | ${[{ ai: 5, bi: 5, c: 1007 /* swap ei -> ie */, p: 1020 }]}
    `(
        'calcSwapCosts with $defs.0 $defs.1 "$wordA"@$ai, "$wordB"@$bi',
        ({
            defs,
            wordA,
            ai,
            wordB,
            bi,
            expected,
        }: {
            defs: SuggestionCostMapDef[];
            wordA: string;
            wordB: string;
            ai: number;
            bi: number;
            expected: CostPosition[];
        }) => {
            const map = createWeightMap(...defs);
            const results = [...map.calcSwapCosts({ a: wordA, b: wordB, ai, bi, c: 1000, p: 1000 })];
            expected.forEach((p) => {
                (p.a = p.a ?? wordA), (p.b = p.b ?? wordB);
            });
            expect(results).toEqual(expect.arrayContaining(expected));
            expect(results).toHaveLength(expected.length);
        }
    );

    test.each`
        defs                                                 | wordA   | wordB   | expected
        ${[defRep('ae', 9), defRep('ei', 7)]}                | ${'a'}  | ${'e'}  | ${9}
        ${[defRep('ae', 9), defRep('ei', 7)]}                | ${'a'}  | ${'i'}  | ${undefined}
        ${[defRep('o(oo)(oh)', 9), defRep('o(oo)(ooo)', 7)]} | ${'oo'} | ${'o'}  | ${7}
        ${[defRep('o(oo)(oh)', 9), defRep('o(oo)(ooo)', 7)]} | ${'oo'} | ${'oh'} | ${9}
    `('calcSwapCosts with $defs.0 $defs.1 "$wordA", "$wordB"', ({ defs, wordA, wordB, expected }) => {
        const map = createWeightMap(...defs);
        const results = lookupReplaceCost(map, wordA, wordB);
        expect(results).toEqual(expected);
    });

    test.each`
        def                                         | expected
        ${defIns('(+A)(+B)', 50, { penalty: 150 })} | ${defIns(sep('(|A)(|B)'), 50, { penalty: 150 })}
        ${defIns('abc', 95)}                        | ${defIns('abc', 95)}
    `('normalizeDef for compound separators $def', ({ def, expected }) => {
        expect(normalizeDef(def)).toEqual(expected);
    });

    test.each`
        adjustments                                                   | word              | expected
        ${[]}                                                         | ${'hello'}        | ${0}
        ${[adj('case-change', /\p{Ll}∙\p{Lu}|\p{Lu}∙\p{Ll}/gu, 500)]} | ${'hello∙There'}  | ${500}
        ${[adj('case-change', /\p{Ll}∙\p{Lu}|\p{Lu}∙\p{Ll}/gu, 500)]} | ${'WORK∙ed'}      | ${500}
        ${[adj('case-change', /\p{Ll}∙\p{Lu}|\p{Lu}∙\p{Ll}/gu, 500)]} | ${'WORK∙ed∙Fine'} | ${1000}
        ${[adj('case-change', /\p{Ll}∙\p{Lu}|\p{Lu}∙\p{Ll}/u, 500)]}  | ${'WORK∙ed∙Fine'} | ${500}
    `('calcAdjustment $adjustments $word', ({ adjustments, word, expected }) => {
        const w = createWeightMap();
        addAdjustment(w, ...adjustments);
        expect(w.calcAdjustment(word)).toEqual(expected);
    });
});

function sep(s: string): string {
    return s.replace(/[|+]/g, DEFAULT_COMPOUNDED_WORD_SEPARATOR);
}

// function mo(...opts: Partial<SuggestionCostMapDef>[]): Partial<SuggestionCostMapDef> {
//     return mergeOps(opts);
// }

function penalty(penalty: number): Partial<SuggestionCostMapDef> {
    return { penalty };
}

function defIns(map: string, insDel: number, ...opts: Partial<SuggestionCostMapDef>[]): SuggestionCostMapDef {
    return { ...mergeOps(opts), map, insDel };
}

function defRep(map: string, replace: number, ...opts: Partial<SuggestionCostMapDef>[]): SuggestionCostMapDef {
    return { ...mergeOps(opts), map, replace };
}

function defSwap(map: string, swap: number, ...opts: Partial<SuggestionCostMapDef>[]): SuggestionCostMapDef {
    return { ...mergeOps(opts), map, swap };
}

function mergeOps(opts: Partial<SuggestionCostMapDef>[]): Partial<SuggestionCostMapDef> {
    return opts.reduce((acc, opt) => ({ ...acc, ...opt }), {} as Partial<SuggestionCostMapDef>);
}

function adj(id: string, regexp: RegExp, penalty: number): PenaltyAdjustment {
    return { id, regexp, penalty };
}
