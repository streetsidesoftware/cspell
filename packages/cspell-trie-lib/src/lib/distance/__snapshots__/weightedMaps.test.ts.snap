// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Validate weightedMaps > buildWeightMap pp { insDel: 11, swap: 4, map: '(ei)(ie)e', replace: 3 }, { map: 'aeiou', replace: 4, insDel: 7 }, { map: 'aeiou', replace: 4, insDel: 7 } 1`] = `
"InsDel:
  (a) = 7
  (e) = 7
  (ei) = 11
  (i) = 7
  (ie) = 11
  (o) = 7
  (u) = 7

Replace:
  (a) -> (e) = 4
  (a) -> (i) = 4
  (a) -> (o) = 4
  (a) -> (u) = 4
  (e) -> (a) = 4
  (e) -> (ei) = 3
  (e) -> (i) = 1
  (e) -> (ie) = 3
  (e) -> (o) = 1
  (e) -> (u) = 4
  (ei) -> (e) = 3
  (ei) -> (ie) = 3
  (i) -> (a) = 4
  (i) -> (e) = 1
  (i) -> (o) = 1
  (i) -> (u) = 4
  (ie) -> (e) = 3
  (ie) -> (ei) = 3
  (o) -> (a) = 4
  (o) -> (e) = 1
  (o) -> (i) = 1
  (o) -> (u) = 4
  (u) -> (a) = 4
  (u) -> (e) = 4
  (u) -> (i) = 4
  (u) -> (o) = 4

Swap:
  (e) <-> (ei) = 4
  (e) <-> (i) = 3
  (e) <-> (ie) = 4
  (e) <-> (o) = 3
  (ei) <-> (e) = 4
  (ei) <-> (ie) = 4
  (i) <-> (e) = 3
  (i) <-> (o) = 3
  (ie) <-> (e) = 4
  (ie) <-> (ei) = 4
  (o) <-> (e) = 3
  (o) <-> (i) = 3
"
`;

exports[`Validate weightedMaps > buildWeightMap pp { map: '(ei)(ie)', replace: 3 }, undefined, undefined 1`] = `
"InsDel:

Replace:
  (ei) -> (ie) = 3
  (ie) -> (ei) = 3

Swap:
"
`;

exports[`Validate weightedMaps > buildWeightMap pp { map: 'ab', insDel: 3 }, { map: 'bc', insDel: 2 }, { map: 'bc', insDel: 2 } 1`] = `
"InsDel:
  (a) = 3
  (b) = 2
  (c) = 2

Replace:

Swap:
"
`;

exports[`Validate weightedMaps > buildWeightMap pp { map: 'ab', insDel: 3 }, undefined, undefined 1`] = `
"InsDel:
  (a) = 3
  (b) = 3

Replace:

Swap:
"
`;

exports[`Validate weightedMaps > buildWeightMap pp { map: 'ab', replace: 3 }, undefined, undefined 1`] = `
"InsDel:

Replace:
  (a) -> (b) = 3
  (b) -> (a) = 3

Swap:
"
`;

exports[`Validate weightedMaps > buildWeightMap pp { map: 'ab', swap: 3 }, undefined, undefined 1`] = `
"InsDel:

Replace:

Swap:
  (a) <-> (b) = 3
  (b) <-> (a) = 3
"
`;

exports[`Validate weightedMaps > buildWeightMap pp undefined, undefined, undefined 1`] = `
"InsDel:

Replace:

Swap:
"
`;
